File: src/lib/api/actuaries.ts
import { api } from './client'
import type { v1ActuaryInfo } from './generated/models/v1ActuaryInfo'
import type { v1ActuaryType } from './generated/models/v1ActuaryType'
import type { v1ListActuariesResponse } from './generated/models/v1ListActuariesResponse'
import type { v1RunDailyResetActuariesResponse } from './generated/models/v1RunDailyResetActuariesResponse'
import type { TradingServiceUpsertActuaryInfoBody } from './generated/models/TradingServiceUpsertActuaryInfoBody'

export type ActuaryInfo = v1ActuaryInfo

export interface ListActuariesArgs {
  emailQuery?: string
  nameQuery?: string
  type?: v1ActuaryType
  page?: number
  pageSize?: number
}

export async function listActuaries(args: ListActuariesArgs = {}): Promise<v1ListActuariesResponse> {
  const { data } = await api.get<v1ListActuariesResponse>('/v1/actuaries', { params: args })
  return data
}

export async function getActuaryInfo(employeeId: string): Promise<v1ActuaryInfo> {
  const { data } = await api.get<v1ActuaryInfo>(`/v1/actuaries/${encodeURIComponent(employeeId)}`)
  return data
}

// upsertActuary creates or updates the actuary_info row attached to
// an employee. Supervisors are forced to dailyLimit=0/needApproval=false
// server-side per spec p.38; pass them in anyway and let the server
// normalise.
export async function upsertActuary(
  employeeId: string,
  body: TradingServiceUpsertActuaryInfoBody,
): Promise<v1ActuaryInfo> {
  const { data } = await api.put<v1ActuaryInfo>(
    `/v1/actuaries/${encodeURIComponent(employeeId)}`,
    body,
  )
  return data
}

export async function updateActuaryLimit(employeeId: string, dailyLimitRsd: string): Promise<v1ActuaryInfo> {
  const { data } = await api.patch<v1ActuaryInfo>(
    `/v1/actuaries/${encodeURIComponent(employeeId)}/limit`,
    { dailyLimit: dailyLimitRsd },
  )
  return data
}

export async function setActuaryNeedApproval(employeeId: string, needApproval: boolean): Promise<v1ActuaryInfo> {
  const { data } = await api.patch<v1ActuaryInfo>(
    `/v1/actuaries/${encodeURIComponent(employeeId)}/need-approval`,
    { needApproval },
  )
  return data
}

export async function resetActuaryUsedLimit(employeeId: string): Promise<v1ActuaryInfo> {
  const { data } = await api.post<v1ActuaryInfo>(
    `/v1/actuaries/${encodeURIComponent(employeeId)}/used-limit/reset`,
    {},
  )
  return data
}

// runActuaryResetJob fires the daily-23:59-Belgrade cron manually.
// Useful for tests + manual reset; in production the cron runs
// itself.
export async function runActuaryResetJob(): Promise<v1RunDailyResetActuariesResponse> {
  const { data } = await api.post<v1RunDailyResetActuariesResponse>('/v1/actuaries/reset-job', {})
  return data
}

File: banka-frontend-enhanced/Banka-3-Backend-rewrite/services/trading/internal/service/actuaries.go
package service

import (
	"context"

	"github.com/RAF-SI-2025/Banka-3-Backend/pkg/apperr"
	"github.com/RAF-SI-2025/Banka-3-Backend/pkg/auth"
	"github.com/RAF-SI-2025/Banka-3-Backend/pkg/money"
	"github.com/RAF-SI-2025/Banka-3-Backend/services/trading/internal/domain"
)

// UpsertActuaryInfoInput carries the supervisor-side promotion of an
// employee to actuary status. The user service is expected to grant
// the matching pkg/permissions Actuary + ActuarySupervisor /
// ActuaryAgent set as a separate step (admin-only); this RPC manages
// only the trading.actuary_info row.
type UpsertActuaryInfoInput struct {
	EmployeeID   string
	Type         domain.ActuaryType
	DailyLimit   string // RSD; ignored for supervisors (forced to "0")
	NeedApproval bool   // ignored for supervisors (forced to false)
}

// UpsertActuaryInfo creates or updates a trading.actuary_info row.
// Supervisors are forced to (limit=0, need_approval=false) per spec
// p.38 ("Supervizor nema limit").
func (s *Service) UpsertActuaryInfo(ctx context.Context, in UpsertActuaryInfoInput) (*domain.ActuaryInfo, error) {
	if _, err := s.requireSupervisor(ctx); err != nil {
		return nil, err
	}
	if in.EmployeeID == "" {
		return nil, apperr.Validation("employee_id is required")
	}
	if in.Type != domain.ActuarySupervisor && in.Type != domain.ActuaryAgent {
		return nil, apperr.Validation("type must be supervisor or agent")
	}

	limit := in.DailyLimit
	need := in.NeedApproval
	if in.Type == domain.ActuarySupervisor {
		limit = "0"
		need = false
	}
	if err := validateNonNegativeAmount(limit); err != nil {
		return nil, err
	}

	return s.Store.UpsertActuaryInfo(ctx, &domain.ActuaryInfo{
		EmployeeID:   in.EmployeeID,
		Type:         in.Type,
		DailyLimit:   limit,
		NeedApproval: need,
	})
}

// GetActuaryInfo returns one actuary record. Visibility:
//   - supervisors / admins see anyone;
//   - the actuary themselves can read their own row (so the FE can
//     render their limit on the trading portal).
func (s *Service) GetActuaryInfo(ctx context.Context, employeeID string) (*domain.ActuaryInfo, error) {
	p, err := s.requirePrincipal(ctx)
	if err != nil {
		return nil, err
	}
	if p.UserID != employeeID {
		if _, err := s.requireSupervisor(ctx); err != nil {
			return nil, err
		}
	}
	return s.Store.GetActuaryInfo(ctx, employeeID)
}

// ListActuaries returns the agents (or supervisors / both) the
// supervisor portal needs. Email/name filters are passed through —
// they're not applied here because actuary_info lacks those columns;
// the gateway / FE composes the list with the user service's people
// directory.
func (s *Service) ListActuaries(ctx context.Context, t domain.ActuaryType, page, pageSize int) ([]*domain.ActuaryInfo, int64, error) {
	if _, err := s.requireSupervisor(ctx); err != nil {
		return nil, 0, err
	}
	return s.Store.ListActuaries(ctx, t, page, pageSize)
}

// UpdateActuaryLimit sets the daily limit. Supervisors only; supervisor
// targets are rejected (their limit is meaningless).
func (s *Service) UpdateActuaryLimit(ctx context.Context, employeeID, dailyLimit string) (*domain.ActuaryInfo, error) {
	if _, err := s.requireSupervisor(ctx); err != nil {
		return nil, err
	}
	if err := validateNonNegativeAmount(dailyLimit); err != nil {
		return nil, err
	}
	cur, err := s.Store.GetActuaryInfo(ctx, employeeID)
	if err != nil {
		return nil, err
	}
	if cur.Type == domain.ActuarySupervisor {
		return nil, apperr.FailedPrecondition("supervizor nema limit")
	}
	// Reject limits below the agent's current used_limit. Otherwise the
	// supervisor can silently put an agent over their cap, which
	// blocks every subsequent order until the cron resets at 23:59.
	newLimit, _ := money.Parse(dailyLimit)
	usedLimit, err := money.Parse(cur.UsedLimit)
	if err != nil {
		return nil, apperr.Validation("used_limit on actuary record is malformed")
	}
	if money.Cmp(newLimit, usedLimit) < 0 {
		return nil, apperr.FailedPrecondition("novi limit ne sme biti manji od trenutno iskorišćenog limita")
	}
	return s.Store.UpdateActuaryLimit(ctx, employeeID, dailyLimit)
}

// ResetActuaryUsedLimit zeroes used_limit for the given agent.
func (s *Service) ResetActuaryUsedLimit(ctx context.Context, employeeID string) (*domain.ActuaryInfo, error) {
	if _, err := s.requireSupervisor(ctx); err != nil {
		return nil, err
	}
	return s.Store.ResetActuaryUsedLimit(ctx, employeeID)
}

// SetActuaryNeedApproval toggles the per-actuary approval requirement.
// Supervisors only; supervisor targets are rejected.
func (s *Service) SetActuaryNeedApproval(ctx context.Context, employeeID string, need bool) (*domain.ActuaryInfo, error) {
	if _, err := s.requireSupervisor(ctx); err != nil {
		return nil, err
	}
	cur, err := s.Store.GetActuaryInfo(ctx, employeeID)
	if err != nil {
		return nil, err
	}
	if cur.Type == domain.ActuarySupervisor {
		return nil, apperr.FailedPrecondition("supervizor uvek ima need_approval=false")
	}
	return s.Store.SetActuaryNeedApproval(ctx, employeeID, need)
}

// RunDailyResetActuaries zeroes used_limit across every actuary. The
// daily 23:59 (Europe/Belgrade) cron calls this; supervisors can call
// it manually via the same RPC for testing / forced reset (spec p.38
// "supervizor ima opciju da resetuje limit i usedLimit za svakog
// agenta, bilo kada").
func (s *Service) RunDailyResetActuaries(ctx context.Context) (int64, error) {
	// The cron path runs without a request principal (no gRPC auth
	// metadata). Allow that explicitly: if there is a principal it
	// must be a supervisor; if there isn't one, accept it as the
	// service-internal call.
	if _, ok := auth.PrincipalFrom(ctx); ok {
		if _, err := s.requireSupervisor(ctx); err != nil {
			return 0, err
		}
	}
	return s.Store.ResetAllUsedLimits(ctx)
}

// validateNonNegativeAmount sanity-checks decimal-string amounts.
func validateNonNegativeAmount(s string) error {
	r, err := money.Parse(s)
	if err != nil {
		return apperr.Validation(err.Error())
	}
	if !money.IsNonNegative(r) {
		return apperr.Validation("amount must be non-negative")
	}
	return nil
}
